web-api/tests/personalized_documents.rs

```diff
@@ -0,0 +1,177 @@
+}
+
+#[derive(Debug, Deserialize, PartialEq)]
+enum PersonalizedDocumentsResponse {
+    #[serde(rename = "documents")]
+    Documents(Vec<PersonalizedDocumentData>),
+    #[serde(rename = "kind")]
+    Error(PersonalizedDocumentsError),
+}
+
+async fn personalize(
+    client: &Client,
+    ingestion_url: &Url,
+    personalization_url: &Url,
+    published_after: Option<&str>,
+) -> Result<Vec<PersonalizedDocumentData>, Panic> {
+    ingest(client, ingestion_url).await?;
+
+    let mut request = client
+        .get(personalization_url.join("/users/u1/personalized_documents")?)
+        .query(&[("count", "5")]);
+    if let Some(published_after) = published_after {
+        request = request.query(&[("published_after", published_after)]);
+    }
+    let request = request.build()?;
+
+    let error = send_assert_json::<PersonalizedDocumentsResponse>(
+        client,
+        request.try_clone().unwrap(),
+        StatusCode::CONFLICT,
+    )
+    .await;
+    assert_eq!(
+        error,
+        PersonalizedDocumentsResponse::Error(PersonalizedDocumentsError::NotEnoughInteractions),
+    );
+
+    interact(client, personalization_url).await?;
+
+    let documents =
+        send_assert_json::<PersonalizedDocumentsResponse>(client, request, StatusCode::OK).await;
+    assert!(matches!(
+        documents,
+        PersonalizedDocumentsResponse::Documents(_),
+    ));
+    let PersonalizedDocumentsResponse::Documents(documents) = documents else {
+        unreachable!();
+    };
+
+    Ok(documents)
+}
+
+#[tokio::test]
+async fn test_personalization_all_dates() {
+    test_two_apps::<Ingestion, Personalization, _>(
+        unchanged_config,
+        unchanged_config,
+        |client, ingestion_url, personalization_url, _| async move {
+            let documents =
+                personalize(&client, &ingestion_url, &personalization_url, None).await?;
+            assert_eq!(
+                documents
+                    .iter()
+                    .map(|document| document.id.as_str())
+                    .collect::<HashSet<_>>(),
+                ["d1", "d4", "d6", "d7", "d8"].into(),
+            );
+
+            Ok(())
+        },
+    )
+    .await;
+}
+
+#[tokio::test]
+async fn test_personalization_limited_dates() {
+    test_two_apps::<Ingestion, Personalization, _>(
+        unchanged_config,
+        unchanged_config,
+        |client, ingestion_url, personalization_url, _| async move {
+            let documents = personalize(
+                &client,
+                &ingestion_url,
+                &personalization_url,
+                Some("2022-01-01T00:00:00Z"),
+            )
+            .await?;
+            assert_eq!(
+                documents
+                    .iter()
+                    .map(|document| document.id.as_str())
+                    .collect::<HashSet<_>>(),
+                ["d1", "d3", "d4"].into(),
+            );
+
+            Ok(())
+        },
+    )
+    .await;
+}
```

Comment:
:pushpin: We could add a test to check that future documents (e.g. a publication date of days in the future) is never included.

Reply:
i'll modify one of the ingested docs to be in the future, then this is also covered by the tests.